pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,336 @@
+		}
+	}
+
+	for pod := range podsForEviction {
+		node, ok := nodeMap[pod.Spec.NodeName]
+		if !ok {
+			klog.ErrorS(err, "Unknown nodeName for pod", "node", pod.Spec.NodeName, "pod", klog.KObj(pod))
+		}
+		if _, err := podEvictor.EvictPod(ctx, pod, node, "PodTopologySpread"); err != nil && !evictable.IsEvictable(pod) {
+			klog.ErrorS(err, "Error evicting pod", "pod", klog.KObj(pod))
+			break
+		}
+	}
+}
+
+// balanceDomains determines how many pods (minimum) should be evicted from large domains to achieve an ideal balance within maxSkew
+// To actually determine how many pods need to be moved, we sort the topology domains in ascending length
+// [2, 5, 3, 8, 5, 7]
+//
+// Would end up like:
+// [2, 3, 5, 5, 7, 8]
+//
+// We then start at i=[0] and j=[len(list)-1] and compare the 2 topology sizes.
+// If the diff of the size of the domains is more than the maxSkew, we will move up to half that skew,
+// or the available pods from the higher domain, or the number required to bring the smaller domain up to the average,
+// whichever number is less.
+//
+// (Note, we will only move as many pods from a domain as possible without bringing it below the ideal average,
+//  and we will not bring any smaller domain above the average)
+// If the diff is within the skew, we move to the next highest domain.
+// If the higher domain can't give any more without falling below the average, we move to the next lowest "high" domain
+//
+// Following this, the above topology domains end up "sorted" as:
+// [5, 5, 5, 5, 5, 5]
+// (assuming even distribution by the scheduler of the evicted pods)
+func balanceDomains(podsForEviction map[*v1.Pod]struct{}, constraint v1.TopologySpreadConstraint, constraintTopologies map[topologyPair][]*v1.Pod, sumPods float64, isEvictable func(*v1.Pod) bool) {
+	idealAvg := sumPods / float64(len(constraintTopologies))
+	sortedDomains := sortDomains(constraintTopologies, isEvictable)
+	// i is the index for belowOrEqualAvg
+	// j is the index for aboveAvg
+	i := 0
+	j := len(sortedDomains) - 1
+	for i < j {
+		// if j has no more to give without falling below the ideal average, move to next aboveAvg
+		if float64(len(sortedDomains[j].pods)) < idealAvg {
+			j--
+		}
+
+		// skew = diff - maxSkew
+		skew := float64(len(sortedDomains[j].pods) - len(sortedDomains[i].pods))
+
+		// if i and j are within the maxSkew of each other, move to next belowOrEqualAvg
+		if int32(skew) <= constraint.MaxSkew {
+			i++
+			continue
+		}
+
+		// the most that can be given from aboveAvg is:
+		// 1. up to half the distance between them
+		// 2. how many it has remaining without falling below the average, or
+		// 3. how many can be added without bringing the smaller domain above the average,
+		// whichever is less
+		// (This is the basic principle of keeping all sizes within ~skew of the average)
+		aboveAvg := math.Ceil(float64(len(sortedDomains[j].pods)) - idealAvg)
```

Comment:
If `idealAvg` is 4.5 and `len(sortedDomains[j].pods` is 5, `aboveAvg` is 1 (0.5 rounded up).
Following what's said in `// 2. how many it has remaining without falling below the average, or`, `1` is remaining though if you remove that one from `sortedDomains[j].pods`, you will get bellow `idealAvg`. Is not the computation  contradicting the comment? The same for `belowAvg` computation.